core/domain/stats_jobs_one_off.py

```diff
@@ -1275,3 +1275,71 @@ def reduce(key, stringified_values):
                     'all: %s sum:%s' % (
                         key, state_name, all_state_hit[state_name],
                         sum_state_hit[state_name]),)
+
+
+class RegenerateMissingStatsModels(jobs.BaseMapReduceOneOffJobManager):
+    """A one-off job to regenerate stats models which were missing due to
+    incorrect handling of exploration reverts. If a model is missing at version
+    x, we will regenerate all models from version x-1 till the max version of
+    the exploration.
+    """
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [exp_models.ExplorationModel]
+
+    @staticmethod
+    def map(exp):
+        all_models = (
+            stats_models.ExplorationStatsModel.get_multi_stats_models(
+                [exp_domain.ExpVersionReference(exp.id, version)
+                 for version in range(1, exp.version + 1)]))
+        first_missing_version = -1
+        for version, model in enumerate(all_models):
+            if model is None:
+                first_missing_version = version + 1
+                break
+
+        if first_missing_version == -1:
+            yield ('No change', exp.id)
+            return
+
+        if first_missing_version == 1:
+            yield ('Missing model at version 1', exp.id)
+            return
```

Comment:
Are you sure you want to return here? Isn't there still valuable data we could gather by checking the rest of the models?

Reply:
Actually yes, I think we need to return. So with the issue we discovered, if a particular version is invalid or buggy, all subsequent versions have to be discarded. AFAICT, the stats for version x+1 just builds on the already known stats for version x. Hence, just knowing which is the first version with an issue should be enough, and we would need to subsequently regenerate all starting from this version.